File: wire/src/agent_workflow.rs
use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;

/// Submit a task to an agent or workflow (`AGDX_AGENT_SUBMIT`). `agent_id` names
/// the target, `run_id` lets the caller assign the run id (else the backend
/// mints one), `params` is scalar control, and `input` is the opaque task body.
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct AgentSubmit {
    pub v: u32,
    pub agent_id: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub run_id: Option<String>,
    #[serde(default, skip_serializing_if = "BTreeMap::is_empty")]
    pub params: BTreeMap<String, String>,
    #[serde(
        default,
        skip_serializing_if = "Option::is_none",
        with = "crate::encoding::opt_bin_bytes"
    )]
    pub input: Option<Vec<u8>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub budget: Option<RunBudget>,
}

/// A per-run resource ceiling carried on submit: caps across independent
/// dimensions, each optional (absent is unbounded on that dimension). A run that
/// crosses any cap is failed with a budget reason.
#[derive(Clone, Debug, Default, PartialEq, Serialize, Deserialize)]
pub struct RunBudget {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub max_events: Option<u64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub max_model_calls: Option<u64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub max_tool_calls: Option<u64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub max_patches: Option<u64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub max_depth: Option<u32>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub max_wall_clock_micros: Option<u64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub max_cost_usd: Option<f64>,
}

/// Cancel a run (`AGDX_AGENT_CANCEL`).
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct AgentCancel {
    pub v: u32,
    pub run_id: String,
}

/// Read a run's status (`AGDX_AGENT_STATUS`).
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct AgentStatusReq {
    pub v: u32,
    pub run_id: String,
}

/// List runs (`AGDX_AGENT_LIST`), filtered and paged. Every field but `v` is
/// optional: an empty request lists the caller's runs from the newest. `limit`
/// is clamped to [`MAX_PAGE_SIZE`](crate::limits::MAX_PAGE_SIZE) and `cursor`
/// is the opaque continuation from the previous page, the kv scan pattern.
#[derive(Clone, Debug, Default, Serialize, Deserialize)]
pub struct AgentList {
    pub v: u32,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub agent_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub state: Option<AgentRunState>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub limit: Option<u32>,
    #[serde(
        default,
        skip_serializing_if = "Option::is_none",
        with = "crate::encoding::opt_bin_bytes"
    )]
    pub cursor: Option<Vec<u8>>,
}

/// A run's lifecycle state. A pinned snake-case vocabulary, additive. The
/// strum derives keep the display, parse, and static-str spellings identical
/// to the serde one by construction (`ForkKind` and `ContentType` set the
/// same pattern).
#[derive(
    Clone,
    Copy,
    Debug,
    Default,
    PartialEq,
    Eq,
    Serialize,
    Deserialize,
    strum::Display,
    strum::EnumString,
    strum::IntoStaticStr,
    strum::VariantArray,
)]
#[serde(rename_all = "snake_case")]
#[strum(serialize_all = "snake_case")]
#[non_exhaustive]
pub enum AgentRunState {
    #[default]
    Submitted,
    Running,
    Completed,
    Cancelled,
    Failed,
}

impl AgentRunState {
    /// The pinned snake-case word, the same spelling serde uses, for query
    /// strings and display.
    pub fn as_str(self) -> &'static str {
        self.into()
    }

    /// Whether the run can never leave this state (the fold refuses to move a
    /// terminal run backward).
    pub fn is_terminal(self) -> bool {
        matches!(
            self,
            AgentRunState::Completed | AgentRunState::Cancelled | AgentRunState::Failed
        )
    }
}

/// A run's metadata, returned by submit, status, and list. `updated_at_micros`
/// is the time of the last state mark (equal to `created_at_micros` until one
/// lands), and `detail` is the terminal summary (an error message on `failed`,
/// absent when clean), so a console renders a run without a second read.
/// `cancel_requested` is the recorded cancel intent, not a state: the engine
/// observes it at its next step boundary and routes it into its own
/// cancellation path, and the state moves only when the engine reports it.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct AgentRunInfo {
    pub run_id: String,
    pub agent_id: String,
    pub user_id: u32,
    pub state: AgentRunState,
    pub created_at_micros: u64,
    pub updated_at_micros: u64,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub detail: Option<String>,
    #[serde(default, skip_serializing_if = "std::ops::Not::not")]
    pub cancel_requested: bool,
}

/// One page of runs: the rows plus the opaque cursor for the next page, absent
/// on the last one.
#[derive(Clone, Debug, Default, Serialize, Deserialize)]
pub struct RunPage {
    pub runs: Vec<AgentRunInfo>,
    #[serde(
        default,
        skip_serializing_if = "Option::is_none",
        with = "crate::encoding::opt_bin_bytes"
    )]
    pub cursor: Option<Vec<u8>>,
}

/// The result of an agent or workflow control op: `Ok` with the outcome, or
/// `Err` with a failure.
#[derive(Clone, Debug, Serialize, Deserialize)]
#[non_exhaustive]
pub enum AgentReply {
    Ok(AgentOutcome),
    Err(AgentError),
}

/// The successful outcome of an agent or workflow command, shaped per op.
#[derive(Clone, Debug, Serialize, Deserialize)]
#[non_exhaustive]
pub enum AgentOutcome {
    Submitted(AgentRunInfo),
    Cancelled(AgentRunInfo),
    Status(AgentRunInfo),
    List(RunPage),
}

/// Why an agent or workflow control op failed.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize, thiserror::Error)]
#[non_exhaustive]
pub enum AgentError {
    #[error("agent ops not supported: {0}")]
    Unsupported(String),
    #[error("run not found: {0}")]
    NotFound(String),
    #[error("invalid agent request: {0}")]
    Invalid(String),
    #[error("agent backend error: {0}")]
    Backend(String),
    #[error("unsupported agent op version (expected {expected}, got {got})")]
    Version { expected: u32, got: u32 },
    /// This plane does not own the mutation partition for the run.
    #[error("not the partition leader for this run")]
    NotLeader,
}

#[cfg(all(test, feature = "cbor"))]
mod tests {
    use super::*;
    use crate::codes::AGENT_WORKFLOW_OP_VERSION;
    use crate::framing::{decode_named, encode_named};

    #[test]
    fn given_a_submit_when_round_tripped_then_should_decode_unchanged() {
        let submit = AgentSubmit {
            v: AGENT_WORKFLOW_OP_VERSION,
            agent_id: "diagnoser".to_owned(),
            run_id: Some("run-7".to_owned()),
            params: BTreeMap::from([("priority".to_owned(), "high".to_owned())]),
            input: Some(br#"{"incident":"INC-7"}"#.to_vec()),
            budget: None,
        };
        let bytes = encode_named(&submit).expect("encodes");
        let back: AgentSubmit = decode_named(&bytes).expect("decodes");
        assert_eq!(back.agent_id, submit.agent_id);
        assert_eq!(back.run_id, submit.run_id);
        assert_eq!(back.input, submit.input);
    }

    #[test]
    fn given_a_reply_when_round_tripped_then_should_preserve_the_variant() {
        let reply = AgentReply::Ok(AgentOutcome::Status(AgentRunInfo {
            run_id: "run-7".to_owned(),
            agent_id: "diagnoser".to_owned(),
            user_id: 42,
            state: AgentRunState::Running,
            created_at_micros: 1_717_171_717_000_000,
            updated_at_micros: 1_717_171_718_000_000,
            detail: None,
            cancel_requested: false,
        }));
        let bytes = encode_named(&reply).expect("encodes");
        let back: AgentReply = decode_named(&bytes).expect("decodes");
        assert!(matches!(
            back,
            AgentReply::Ok(AgentOutcome::Status(info)) if info.run_id == "run-7"
        ));
    }

    #[test]
    fn given_a_filtered_list_when_round_tripped_then_should_keep_filters_and_cursor() {
        let list = AgentList {
            v: AGENT_WORKFLOW_OP_VERSION,
            agent_id: Some("diagnoser".to_owned()),
            state: Some(AgentRunState::Running),
            limit: Some(25),
            cursor: Some(vec![0x01, 0x02]),
        };
        let bytes = encode_named(&list).expect("encodes");
        let back: AgentList = decode_named(&bytes).expect("decodes");
        assert_eq!(back.agent_id, list.agent_id);
        assert_eq!(back.state, list.state);
        assert_eq!(back.limit, list.limit);
        assert_eq!(back.cursor, list.cursor);
    }

    #[test]
    fn given_run_state_words_when_parsed_then_should_round_trip_through_display() {
        for state in [
            AgentRunState::Submitted,
            AgentRunState::Running,
            AgentRunState::Completed,
            AgentRunState::Cancelled,
            AgentRunState::Failed,
        ] {
            let parsed: AgentRunState = state.as_str().parse().expect("pinned word parses");
            assert_eq!(parsed, state);
        }
        assert!("paused".parse::<AgentRunState>().is_err());
    }

    #[test]
    fn given_an_empty_list_request_when_encoded_then_should_skip_absent_fields() {
        let bare = AgentList {
            v: AGENT_WORKFLOW_OP_VERSION,
            ..AgentList::default()
        };
        let bytes = encode_named(&bare).expect("encodes");
        let back: AgentList = decode_named(&bytes).expect("decodes");
        assert!(back.agent_id.is_none());
        assert!(back.state.is_none());
        assert!(back.limit.is_none());
        assert!(back.cursor.is_none());
    }
}

File: foreign/typescript/src/wire/agent-workflow.ts
import { CodecError } from "../client/errors.js"
import { AGENT_WORKFLOW_OP_VERSION } from "./codes.js"
import { type CborMap, expectMap, expectString, field, singleVariantTag } from "./cbor.js"

export interface RunBudget {
  readonly maxEvents?: bigint
  readonly maxModelCalls?: bigint
  readonly maxToolCalls?: bigint
  readonly maxPatches?: bigint
  readonly maxDepth?: number
  readonly maxWallClockMicros?: bigint
  readonly maxCostUsd?: number
}

export function encodeRunBudget(budget: RunBudget): Map<string, unknown> {
  const map = new Map<string, unknown>()
  if (budget.maxEvents !== undefined) map.set("max_events", budget.maxEvents)
  if (budget.maxModelCalls !== undefined) map.set("max_model_calls", budget.maxModelCalls)
  if (budget.maxToolCalls !== undefined) map.set("max_tool_calls", budget.maxToolCalls)
  if (budget.maxPatches !== undefined) map.set("max_patches", budget.maxPatches)
  if (budget.maxDepth !== undefined) map.set("max_depth", budget.maxDepth)
  if (budget.maxWallClockMicros !== undefined) {
    map.set("max_wall_clock_micros", budget.maxWallClockMicros)
  }
  if (budget.maxCostUsd !== undefined) map.set("max_cost_usd", budget.maxCostUsd)
  return map
}

export function decodeRunBudget(map: CborMap, context: string): RunBudget {
  const maxEvents = field.optionalU64(map, "max_events", context)
  const maxModelCalls = field.optionalU64(map, "max_model_calls", context)
  const maxToolCalls = field.optionalU64(map, "max_tool_calls", context)
  const maxPatches = field.optionalU64(map, "max_patches", context)
  const maxDepth = field.optionalU32(map, "max_depth", context)
  const maxWallClockMicros = field.optionalU64(map, "max_wall_clock_micros", context)
  const maxCostUsd = field.optionalF64(map, "max_cost_usd", context)
  return {
    ...(maxEvents !== undefined ? { maxEvents } : {}),
    ...(maxModelCalls !== undefined ? { maxModelCalls } : {}),
    ...(maxToolCalls !== undefined ? { maxToolCalls } : {}),
    ...(maxPatches !== undefined ? { maxPatches } : {}),
    ...(maxDepth !== undefined ? { maxDepth } : {}),
    ...(maxWallClockMicros !== undefined ? { maxWallClockMicros } : {}),
    ...(maxCostUsd !== undefined ? { maxCostUsd } : {})
  }
}

function encodeStringMap(entries: ReadonlyMap<string, string>): Map<string, unknown> {
  return new Map(entries)
}

function decodeStringMap(map: CborMap, context: string): ReadonlyMap<string, string> {
  const result = new Map<string, string>()
  for (const [key, value] of map) {
    if (typeof key !== "string" || typeof value !== "string") {
      throw new CodecError(`${context} must map strings to strings`, context, "params")
    }
    result.set(key, value)
  }
  return result
}

export interface AgentSubmit {
  readonly agentId: string
  readonly runId?: string
  readonly params: ReadonlyMap<string, string>
  readonly input?: Uint8Array
  readonly budget?: RunBudget
}

export function encodeAgentSubmit(submit: AgentSubmit): Map<string, unknown> {
  const map = new Map<string, unknown>()
  map.set("v", AGENT_WORKFLOW_OP_VERSION)
  map.set("agent_id", submit.agentId)
  if (submit.runId !== undefined) map.set("run_id", submit.runId)
  if (submit.params.size > 0) map.set("params", encodeStringMap(submit.params))
  if (submit.input !== undefined) map.set("input", submit.input)
  if (submit.budget !== undefined) map.set("budget", encodeRunBudget(submit.budget))
  return map
}

export function decodeAgentSubmit(map: CborMap, context: string): AgentSubmit {
  const agentId = field.requiredString(map, "agent_id", context)
  const runId = field.optionalString(map, "run_id", context)
  const paramsMap = field.optionalMap(map, "params", context)
  const input = field.optionalBytes(map, "input", context)
  const budgetMap = field.optionalMap(map, "budget", context)
  return {
    agentId,
    ...(runId !== undefined ? { runId } : {}),
    params: paramsMap !== undefined ? decodeStringMap(paramsMap, `${context}.params`) : new Map(),
    ...(input !== undefined ? { input } : {}),
    ...(budgetMap !== undefined ? { budget: decodeRunBudget(budgetMap, `${context}.budget`) } : {})
  }
}

export interface AgentCancel {
  readonly runId: string
}

export function encodeAgentCancel(cancel: AgentCancel): Map<string, unknown> {
  return new Map<string, unknown>([
    ["v", AGENT_WORKFLOW_OP_VERSION],
    ["run_id", cancel.runId]
  ])
}

export function decodeAgentCancel(map: CborMap, context: string): AgentCancel {
  return { runId: field.requiredString(map, "run_id", context) }
}

export interface AgentStatusReq {
  readonly runId: string
}

export function encodeAgentStatusReq(req: AgentStatusReq): Map<string, unknown> {
  return new Map<string, unknown>([
    ["v", AGENT_WORKFLOW_OP_VERSION],
    ["run_id", req.runId]
  ])
}

export function decodeAgentStatusReq(map: CborMap, context: string): AgentStatusReq {
  return { runId: field.requiredString(map, "run_id", context) }
}

export type AgentRunState = "submitted" | "running" | "completed" | "cancelled" | "failed"

const AGENT_RUN_STATES: ReadonlySet<string> = new Set([
  "submitted",
  "running",
  "completed",
  "cancelled",
  "failed"
])

export function agentRunStateFromWord(word: string, context: string): AgentRunState {
  if (!AGENT_RUN_STATES.has(word)) {
    throw new CodecError(`\`${word}\` is not a recognized agent run state`, context, "state")
  }
  return word as AgentRunState
}

export function agentRunStateIsTerminal(state: AgentRunState): boolean {
  return state === "completed" || state === "cancelled" || state === "failed"
}

export interface AgentRunInfo {
  readonly runId: string
  readonly agentId: string
  readonly userId: number
  readonly state: AgentRunState
  readonly createdAtMicros: bigint
  readonly updatedAtMicros: bigint
  readonly detail?: string
  readonly cancelRequested: boolean
}

export function encodeAgentRunInfo(info: AgentRunInfo): Map<string, unknown> {
  const map = new Map<string, unknown>()
  map.set("run_id", info.runId)
  map.set("agent_id", info.agentId)
  map.set("user_id", info.userId)
  map.set("state", info.state)
  map.set("created_at_micros", info.createdAtMicros)
  map.set("updated_at_micros", info.updatedAtMicros)
  if (info.detail !== undefined) map.set("detail", info.detail)
  if (info.cancelRequested) map.set("cancel_requested", true)
  return map
}

export function decodeAgentRunInfo(map: CborMap, context: string): AgentRunInfo {
  const runId = field.requiredString(map, "run_id", context)
  const agentId = field.requiredString(map, "agent_id", context)
  const userId = field.requiredU32(map, "user_id", context)
  const state = agentRunStateFromWord(field.requiredString(map, "state", context), context)
  const createdAtMicros = field.requiredU64(map, "created_at_micros", context)
  const updatedAtMicros = field.requiredU64(map, "updated_at_micros", context)
  const detail = field.optionalString(map, "detail", context)
  return {
    runId,
    agentId,
    userId,
    state,
    createdAtMicros,
    updatedAtMicros,
    ...(detail !== undefined ? { detail } : {}),
    cancelRequested: map.get("cancel_requested") === true
  }
}

export interface AgentList {
  readonly agentId?: string
  readonly state?: AgentRunState
  readonly limit?: number
  readonly cursor?: Uint8Array
}

export function encodeAgentList(list: AgentList): Map<string, unknown> {
  const map = new Map<string, unknown>()
  map.set("v", AGENT_WORKFLOW_OP_VERSION)
  if (list.agentId !== undefined) map.set("agent_id", list.agentId)
  if (list.state !== undefined) map.set("state", list.state)
  if (list.limit !== undefined) map.set("limit", list.limit)
  if (list.cursor !== undefined) map.set("cursor", list.cursor)
  return map
}

export function decodeAgentList(map: CborMap, context: string): AgentList {
  const agentId = field.optionalString(map, "agent_id", context)
  const state = field.optionalString(map, "state", context)
  const limit = field.optionalU32(map, "limit", context)
  const cursor = field.optionalBytes(map, "cursor", context)
  return {
    ...(agentId !== undefined ? { agentId } : {}),
    ...(state !== undefined ? { state: agentRunStateFromWord(state, context) } : {}),
    ...(limit !== undefined ? { limit } : {}),
    ...(cursor !== undefined ? { cursor } : {})
  }
}

export interface RunPage {
  readonly runs: readonly AgentRunInfo[]
  readonly cursor?: Uint8Array
}

export function encodeRunPage(page: RunPage): Map<string, unknown> {
  const map = new Map<string, unknown>()
  map.set(
    "runs",
    page.runs.map((run) => encodeAgentRunInfo(run))
  )
  if (page.cursor !== undefined) map.set("cursor", page.cursor)
  return map
}

export function decodeRunPage(map: CborMap, context: string): RunPage {
  const runs = field.requiredArray(map, "runs", context, (item, index) =>
    decodeAgentRunInfo(
      expectMap(item, `${context}.runs[${String(index)}]`),
      `${context}.runs[${String(index)}]`
    )
  )
  const cursor = field.optionalBytes(map, "cursor", context)
  return { runs, ...(cursor !== undefined ? { cursor } : {}) }
}

export type AgentOutcome =
  | { readonly kind: "submitted"; readonly run: AgentRunInfo }
  | { readonly kind: "cancelled"; readonly run: AgentRunInfo }
  | { readonly kind: "status"; readonly run: AgentRunInfo }
  | { readonly kind: "list"; readonly page: RunPage }
  | { readonly kind: "unrecognized"; readonly tag: string; readonly value: unknown }

export type AgentWorkflowError =
  | { readonly kind: "unsupported"; readonly message: string }
  | { readonly kind: "notFound"; readonly message: string }
  | { readonly kind: "invalid"; readonly message: string }
  | { readonly kind: "backend"; readonly message: string }
  | { readonly kind: "version"; readonly expected: number; readonly got: number }
  | { readonly kind: "notLeader" }
  | { readonly kind: "unrecognized"; readonly tag: string; readonly value: unknown }

export type AgentReply =
  | { readonly kind: "ok"; readonly outcome: AgentOutcome }
  | { readonly kind: "err"; readonly error: AgentWorkflowError }
  | { readonly kind: "unrecognized"; readonly tag: string; readonly value: unknown }

export function encodeAgentOutcome(outcome: AgentOutcome): Map<string, unknown> {
  switch (outcome.kind) {
    case "submitted":
      return new Map([["Submitted", encodeAgentRunInfo(outcome.run)]])
    case "cancelled":
      return new Map([["Cancelled", encodeAgentRunInfo(outcome.run)]])
    case "status":
      return new Map([["Status", encodeAgentRunInfo(outcome.run)]])
    case "list":
      return new Map([["List", encodeRunPage(outcome.page)]])
    case "unrecognized":
      return new Map([[outcome.tag, outcome.value]])
  }
}

export function decodeAgentOutcome(value: unknown, context: string): AgentOutcome {
  const [tag, inner] = singleVariantTag(value, context)
  switch (tag) {
    case "Submitted":
      return { kind: "submitted", run: decodeAgentRunInfo(expectMap(inner, context), context) }
    case "Cancelled":
      return { kind: "cancelled", run: decodeAgentRunInfo(expectMap(inner, context), context) }
    case "Status":
      return { kind: "status", run: decodeAgentRunInfo(expectMap(inner, context), context) }
    case "List":
      return { kind: "list", page: decodeRunPage(expectMap(inner, context), context) }
    default:
      return { kind: "unrecognized", tag, value: inner }
  }
}

export function encodeAgentWorkflowError(error: AgentWorkflowError): unknown {
  switch (error.kind) {
    case "unsupported":
      return new Map([["Unsupported", error.message]])
    case "notFound":
      return new Map([["NotFound", error.message]])
    case "invalid":
      return new Map([["Invalid", error.message]])
    case "backend":
      return new Map([["Backend", error.message]])
    case "version":
      return new Map([
        [
          "Version",
          new Map<string, unknown>([
            ["expected", error.expected],
            ["got", error.got]
          ])
        ]
      ])
    case "notLeader":
      return "NotLeader"
    case "unrecognized":
      return new Map([[error.tag, error.value]])
  }
}

export function decodeAgentWorkflowError(value: unknown, context: string): AgentWorkflowError {
  if (typeof value === "string") {
    return value === "NotLeader"
      ? { kind: "notLeader" }
      : { kind: "unrecognized", tag: value, value: undefined }
  }
  const [tag, inner] = singleVariantTag(value, context)
  switch (tag) {
    case "Unsupported":
      return { kind: "unsupported", message: expectString(inner, context) }
    case "NotFound":
      return { kind: "notFound", message: expectString(inner, context) }
    case "Invalid":
      return { kind: "invalid", message: expectString(inner, context) }
    case "Backend":
      return { kind: "backend", message: expectString(inner, context) }
    case "Version": {
      const versionMap = expectMap(inner, context)
      return {
        kind: "version",
        expected: field.requiredU32(versionMap, "expected", context),
        got: field.requiredU32(versionMap, "got", context)
      }
    }
    default:
      return { kind: "unrecognized", tag, value: inner }
  }
}

export function encodeAgentReply(reply: AgentReply): Map<string, unknown> {
  switch (reply.kind) {
    case "ok":
      return new Map([["Ok", encodeAgentOutcome(reply.outcome)]])
    case "err":
      return new Map([["Err", encodeAgentWorkflowError(reply.error)]])
    case "unrecognized":
      return new Map([[reply.tag, reply.value]])
  }
}

export function decodeAgentReply(value: unknown, context: string): AgentReply {
  const [tag, inner] = singleVariantTag(value, context)
  switch (tag) {
    case "Ok":
      return { kind: "ok", outcome: decodeAgentOutcome(inner, context) }
    case "Err":
      return { kind: "err", error: decodeAgentWorkflowError(inner, context) }
    default:
      return { kind: "unrecognized", tag, value: inner }
  }
}
